lighthouse-core/config/config-plugin.js

```diff
@@ -0,0 +1,176 @@
+      }
+
+      if (typeof id !== 'string') {
+        throw new Error(`${fullPluginName} has an invalid auditRef id.`);
+      }
+      if (typeof weight !== 'number') {
+        throw new Error(`${fullPluginName} has an invalid auditRef weight.`);
+      }
+
+      return {
+        id,
+        weight,
+      };
+    });
+  }
+
+  /**
+   * Extract and validate the category added by the plugin.
+   * @param {unknown} categoryJson
+   * @param {string} fullPluginName
+   * @return {LH.Config.CategoryJson}
+   */
+  static _parseCategory(categoryJson, fullPluginName) {
+    if (!isObjectOfUnknownProperties(categoryJson)) {
+      throw new Error(`${fullPluginName} has no valid category.`);
+    }
+
+    const {
+      title,
+      description,
+      manualDescription,
+      auditRefs: auditRefsJson,
+      ...invalidRest
+    } = categoryJson;
+
+    const invalidKeys = Object.keys(invalidRest);
+    if (invalidKeys.length > 0) {
+      const keys = invalidKeys.join(', ');
+      throw new Error(`${fullPluginName} has invalid category properties [${keys}].`);
+    }
+
+    if (typeof title !== 'string') {
+      throw new Error(`${fullPluginName} has an invalid category tile.`);
+    }
+    if (typeof description !== 'string' && typeof description !== 'undefined') {
+      throw new Error(`${fullPluginName} has an invalid category description.`);
+    }
+    if (typeof manualDescription !== 'string' && typeof manualDescription !== 'undefined') {
+      throw new Error(`${fullPluginName} has an invalid category manualDescription.`);
+    }
+    const auditRefs = ConfigPlugin._parseAuditRefsList(auditRefsJson, fullPluginName);
+
+    return {
+      title,
+      auditRefs,
+      description: description,
+      manualDescription: manualDescription,
+    };
+  }
+
+  /**
+   * Extracts and validates a ConfigJson from the provided plugin input, throwing
+   * if it deviates from the expected object shape.
+   * @param {unknown} pluginJson
+   * @param {string} pluginName
+   * @return {LH.Config.Json}
+   */
+  static parsePlugin(pluginJson, pluginName) {
+    const fullName = `lighthouse-plugin-${pluginName}`;
+
+    // Clone to prevent modifications or original and to deactivate any live properties.
+    pluginJson = JSON.parse(JSON.stringify(pluginJson));
+    if (!isObjectOfUnknownProperties(pluginJson)) {
+      throw new Error(`${fullName} is not defined as an object.`);
+    }
+
+    const {
+      audits: pluginAuditsJson,
+      category: pluginCategoryJson,
+      ...invalidRest
+    } = pluginJson;
+
+    // Assert no other properties found.
+    const invalidKeys = Object.keys(invalidRest);
+    if (invalidKeys.length > 0) {
+      throw new Error(`${fullName} has invalid properties [${invalidKeys.join(', ')}].`);
+    }
+
+    return {
+      audits: ConfigPlugin._parseAuditsList(pluginAuditsJson, fullName),
+      categories: {
+        [pluginName]: ConfigPlugin._parseCategory(pluginCategoryJson, fullName),
```

Comment:
we should probably check that this isn't one of the built-in category ids

otherwise `lighthouse-plugin-performance` is gonna get free reign ;)

Reply:
> we should probably check that this isn't one of the built-in category ids

gooood call. It's not completely clear which ones to ban, so I just did the default categories and any of the categories in the config being used (in case they've already added more categories, we don't want the plugin to merge in).

Maybe we want to have a set of reserved words, though? I can see `lighthouse-plugin-security` being snapped up, for instance, but we might want that later.

Reply:
> Maybe we want to have a set of reserved words, though? I can see lighthouse-plugin-security being snapped up, for instance, but we might want that later.

Now that I say that, this situation seems terrible :)

These are just ids in the LHR, not visible in the report. Maybe we should just have the category ID be the full `lighthouse-plugin-pluginName`? It would still be well known and obvious to the user since they'd have to `npm install` it or name a folder that or whatever.

Reply:
Yeah that sounds good 👍

Reply:
>  Maybe we should just have the category ID be the full lighthouse-plugin-pluginName

going to do this.

@paulirish brings up that after that's done, the only real use of the short name is in the config. Considering you have to use the fully qualified name everywhere else (creation time, `package.json`, getting the result from the LHR), we might as well use the full name in the config as well. You only type it once(ish) and you probably have less surprise trying to get from a reference in a config to the source on `npm` or in `node_modules/`.

So I'm just going to get rid of the short version.